src/Models/BaseElement.php

```diff
@@ -714,17 +714,29 @@ public function Link($action = null)
 
     /**
      * @param string|null $action
-     * @return string
+     * @return string|null
      * @throws \Psr\Container\NotFoundExceptionInterface
      * @throws \SilverStripe\ORM\ValidationException
      */
     public function PreviewLink($action = null)
     {
-        $action = $action . '?ElementalPreview=' . mt_rand();
-        $link = $this->Link($action);
-        $this->extend('updatePreviewLink', $link);
+        if ($page = $this->getPage()) {
+            $link = null;
+            if (ClassInfo::hasMethod($page, 'Link')) {
+                $link = $page->Link($action);
+            } else if ($page instanceof CMSPreviewable) {
+                $link = $page->PreviewLink($action);
+            }
+            if ($link) {
+                // The ElementalPreview getvar is used in ElementalPageExtension
+                // The anchor must be at the end of the URL to function correctly
+                $link .= '?ElementalPreview=' . mt_rand() . '#' . $this->getAnchor();
+                $this->extend('updatePreviewLink', $link);
```

Comment:
Probably, we should provide users with ability to extend PreviewLink even if link is NULL.